계산기 [STEP2] Hisop

Hi-sop opened the pull request:
@juun97
안녕하세요! STEP2 PR 보냅니다.

## 고민했던 점
UML을 보고 의도를 해석하는 것이 쉽지 않았습니다. 
- split의 extension은 왜 만들어야 했는지 아직 잘 모르겠습니다. 기존 split과 다른 기능을 만들어보려고 했으나 필요성을 느끼지 못했습니다.
따라서 리턴값을 subString에서 String으로 변경하는 코드만 추가해주었습니다.
- componentsByOperator가 Operator에 따라 components를 분리하는 기능만 하도록 만들어보려고 했습니다.
처음엔 extension split의 기능을 이용하여 '개행으로 구분된 components들의 String', 'Operator를 모은 String'을 리턴하려 했으나 비효율적인것 같아 현재 코드로 변경했습니다.

## 조언을 얻고 싶었던 점
- 음수 등 여러 케이스들에 대한 대응을 거의 하지 않았다는 생각이 듭니다. 계산기라는 특성상 들어올 일이 거의 없을거란 생각이 들긴 합니다만 생각하지 않고 넘어가도 괜찮은걸까요
- 현재 제 유닛테스트는 다른 객체의 메서드를 사용해 테스트를 진행하고 있는데 수정하지 않아도 괜찮은 부분인가요?

Commit:
feat: Step2 UML에따라 타입 선언

Commit:
feat: Operator구현 및 테스트 추가

Commit:
feat: Formula구현 및 테스트 추가

Commit:
feat: ExpressionParser 구현 및 테스트 추가

Extension 별도의 파일로 분리하여 관리

juun97 commented on Calculator/Calculator/ExpressionParser.swift:
```diff
@@ -0,0 +1,36 @@
+
+enum ExpressionParser {
+    static func parse(from input: String) -> Formula {
+        let components = componentsByOperator(from: input).compactMap{ Double($0) }
```
componentsByOperator 메서드의 결과값은 어떤 components 인가요? 좀더 명확한 네이밍이 필요할 것 같습니다.

Hi-sop replied:
NumberElements로 네이밍 변경하였습니다.

juun97 commented on Calculator/Calculator/Extension.swift:
```diff
@@ -0,0 +1,14 @@
+extension Double: CalculateItem { }
+
+extension String {
+    func split(with target: Character) -> [String] {
+        return split(separator: target).map { String($0) }
```
1. split 의 반환타입은 무엇인가요?
2. split 과 components 의 차이점은 무엇인가요?

Hi-sop replied:
1. [subString]
2. components는 문자열로 쪼갤 수 있으며 리턴이 [String]입니다. https://developer.apple.com/documentation/foundation/nsstring/1413214-components참고했습니다.

juun97 commented on Calculator/Calculator/Formula.swift:
```diff
@@ -0,0 +1,27 @@
+        }
+        
+        while let operand = operands.dequeue() {
+            guard let operatorType = operators.dequeue() else {
```
[API Design Guidelines](https://www.swift.org/documentation/api-design-guidelines/)에 따르면 변수명에 Type 이란 단어가 직접적으로 들어가선 안됩니다.

Type 이란 단어는 특정 객체의 Type 그 자체를 의미하기 때문에 다른 네이밍이 필요해 보이네요.

Hi-sop replied:
`operator`로 네이밍 변경하였습니다.

juun97 commented on Calculator/Calculator/Formula.swift:
```diff
@@ -0,0 +1,27 @@
+            return .nan
+        }
+        
+        while let operand = operands.dequeue() {
```
While let binding 을 잘 활용해주셨네요 👍🏻

juun97 commented on Calculator/Calculator/Operator.swift:
```diff
@@ -0,0 +1,42 @@
+//  Created by Hisop on 2023/10/11.
+//
+
+enum Operator: Character, CaseIterable, CalculateItem {
```
CaseIterable 은 어떤 프로토콜인가요?

Hi-sop replied:
.allCases 를 사용하여 모든 케이스를 모은 컬렉션에 접근 가능하게 한다고 이해했습니다.

https://developer.apple.com/documentation/swift/caseiterable

juun97 commented on Calculator/CalculatorTests/FormulaTest.swift:
```diff
@@ -0,0 +1,43 @@
+        sut = nil
+    }
+
+    func test_operands에10_20_5_5_2_operators에add_subtract_divide_multiply_추가한뒤_result의_반환값이10인지() throws {
+        let expectation: Double = 10
+        
+        sut.operands.enqueue(value: 10)
+    
+        sut.operators.enqueue(value: .add)
+        sut.operands.enqueue(value: 20)
+        
+        sut.operands.enqueue(value: 5)
+        sut.operators.enqueue(value: .subtract)
+        
+        sut.operands.enqueue(value: 5)
+        sut.operators.enqueue(value: .divide)
+
+        sut.operands.enqueue(value: 2)
+        sut.operators.enqueue(value: .multiply)
```
operators 와 operands 들의 enqueue 를 각각에 맞게 묶어주거나 순서를 지켜 작성해주시면 보기 좋을 것 같습니다.

Hi-sop replied:
operators와 operands로 각각 묶이도록 순서 변경했습니다.

juun97 reviewed (commented):
안녕하세요 Hisop 스텝2 너무 훌륭하게 수행해주셔서 크게 코멘트 드릴 부분이 없었네요 !

## 고민했던 점
> - split의 extension은 왜 만들어야 했는지 아직 잘 모르겠습니다. 기존 split과 다른 기능을 만들어보려고 했으나 필요성을 느끼지 못했습니다.
따라서 리턴값을 subString에서 String으로 변경하는 코드만 추가해주었습니다.

split 의 경우 반환타입이 subString 의 배열입니다. 하지만 subString 은 그 자체로는 사용이 불가능합니다. 그렇기 때문에 변환과정을 필요로 하게 되는데 split 을 하는 모든 곳에서 중복된 작업을 해야 하기 때문에 extension으로 공통된 작업을 처리를 하고 있는 것입니다.

split 이외에도 components 라는 메서드가 있으니 알아보시면 좋을 것 같습니다.

> - componentsByOperator가 Operator에 따라 components를 분리하는 기능만 하도록 만들어보려고 했습니다.
처음엔 extension split의 기능을 이용하여 '개행으로 구분된 components들의 String', 'Operator를 모은 String'을 리턴하려 했으나 비효율적인것 같아 현재 코드로 변경했습니다.

의도에 맞게 잘 구현해주신 것 같습니다.


## 조언을 얻고 싶었던 점
> - 음수 등 여러 케이스들에 대한 대응을 거의 하지 않았다는 생각이 듭니다. 계산기라는 특성상 들어올 일이 거의 없을거란 생각이 들긴 합니다만 생각하지 않고 넘어가도 괜찮은걸까요

테스트케이스는 모든 상황에 맞춰 대응을 해야 합니다. 만약 미래에 음수에 대한 처리가 필요해질 가능성이 조금이라도 있다면 대비를 해두셔야합니다.

종종 "넘어가도 괜찮다" 고 생각할 수 있는 부분이 미래에 문제를 일으킬 수 있으므로, 프로젝트 목표와 장기적인 유지보수를 고려해 모든 상황에 맞춰 준비를 해두시는게 좋습니다.

> - 현재 제 유닛테스트는 다른 객체의 메서드를 사용해 테스트를 진행하고 있는데 수정하지 않아도 괜찮은 부분인가요?

유닛 테스트를 작성할 때 다른 객체의 메서드를 사용하는 것은 일반적인 접근 방식입니다. 추후에 Test Double 이라는 개념을 배우실 겁니다. 이를 모의 객체(Mock Object) 또는 스텁(Stub)을 사용함으로써 테스트 환경을 제어하고 코드 일관성을 유지할 수 있습니다.

Commit:
refactor: PR코멘트 반영, 네이밍 및 컨벤션 등

componentsByOperator의input.split 제거. 리턴하며 공백을 지우는 형태라 불필요한 코드라 판단함
Operator의 '-'을 'minus sign'으로 변경
FormulaTest의 컨벤션 변경
parse메서드의 네이밍 변경 components -> numberElements
Formula.result 메서드의 네이밍 변경 OperatorType ->
